app/src/main/java/com/example/moviewithfragments/request/MovieApiClient.kt

```diff
@@ -0,0 +1,141 @@
+import androidx.lifecycle.MutableLiveData
+import com.example.moviewithfragments.AppExecutors
+import com.example.moviewithfragments.models.MovieModel
+import com.example.moviewithfragments.response.MovieSearchResponse
+import com.example.moviewithfragments.utils.Credentials
+import retrofit2.Call
+import retrofit2.Response
+import java.io.IOException
+import java.util.concurrent.Future
+import java.util.concurrent.TimeUnit
+
+class MovieApiClient private constructor() {
+
+    var appExecutors = AppExecutors()
+    private val viewModelMovies: MutableLiveData<List<MovieModel>?>
+    var retrieveMoviesRunnable: RetrieveMoviesRunnable? = null
+    val movies: MutableLiveData<List<MovieModel>?>
+        get() = viewModelMovies
+
+    companion object {
+        var instance: MovieApiClient? = null
+            get() {
+                if (field == null) {
+                    field = MovieApiClient()
+                }
+                return field
+            }
+            private set
+    }
+
+    init {
+        viewModelMovies = MutableLiveData()
+    }
+
+    fun searchMoviesApi(query: String, year: Int, pageNumber: Int) {
+
+        if (retrieveMoviesRunnable != null) {
+            retrieveMoviesRunnable = null
+        }
+        val retrieveMoviesRunnable = RetrieveMoviesRunnable(query, year, pageNumber)
+        val retroFitHandler: Future<*> =
+            appExecutors.getInstance()?.getNetworkIO()!!.submit(retrieveMoviesRunnable)
+        appExecutors.getInstance()!!.getNetworkIO()?.schedule(Runnable {
+            @Override
+            fun run() {
+                retroFitHandler.cancel(true)
+
+            }
+        }, 5000, TimeUnit.MILLISECONDS)
+    }
+
+    inner class RetrieveMoviesRunnable(
+        private val query: String,
+        private val year: Int,
+        private val pageNumber: Int
+    ) :
+        Runnable {
+
+        var cancelRequest = false
+        override fun run() {
+            try {
+                if (query.length != 0) {
+                    val response = getMovies(query, pageNumber)?.execute()
+                    responseData(response)
+                } else {
+                    if (year == 0) {
+                        val response = getPopularMovies()?.execute()
+                        responseData(response)
+                    } else {
+                        val response = getLatestMovies(year)?.execute()
+                        responseData(response)
+                    }
+                }
+            } catch (e: IOException) {
+                e.printStackTrace()
+            }
+        }
+
+        fun responseData(response: Response<MovieSearchResponse?>?) {
+            if (cancelRequest) {
+                return
+            }
+            if (response!!.code() == 200) {
+                val list = response.body()!!.movies as java.util.ArrayList<MovieModel>?
+                if (pageNumber == 1) {
+                    viewModelMovies?.postValue(list as List<MovieModel>?)
```

Comment:
we should give the response as callback from apiclient to repository.
Normally LiveData always works on MainThread.